File: JavaScript/5_DataStructures/2_DoublyLinkedList.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
    this.prev = null;
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  printAll() {
    let output = [];
    if (!this.head) return false;
    let node = this.head;
    while (node) {
      output.push(node.val);
      node = node.next;
    }
    return output;
  }

  push(val) {
    const new_node = new Node(val);
    if (!this.head) {
      this.head = new_node;
      this.tail = new_node;
    } else {
      this.tail.next = new_node;
      new_node.prev = this.tail;
      this.tail = new_node;
    }
    this.length++;
    return true;
  }

  pop() {
    if (!this.head) return false;
    let oldTail = this.tail;
    if (this.length === 1) {
      this.head = null;
      this.tail = null;
      this.length = 0;
      return oldTail;
    }
    this.tail = oldTail.prev;
    this.tail.next = null;
    oldTail.prev = null;
    this.length--;
    return oldTail;
  }

  shift(val) {
    const new_node = new Node(val);
    if (!this.head) {
      this.head = new_node;
      this.tail = new_node;
    } else {
      this.head.prev = new_node;
      new_node.next = this.head;
      this.head = new_node;
    }
    this.length++;
    return true;
  }

  unShift() {
    if (!this.head) return false;
    if (this.length === 1) {
      this.head = null;
      this.tail = null;
      this.length = 0;
      return;
    }
    let oldHead = this.head;
    this.head = this.head.next;
    this.head.prev = null;
    oldHead.next = null;
    this.length--;
    return oldHead;
  }

  get(index) {
    if (index < 0 || index >= this.length) return false;
    let node;
    if (index <= this.length / 2) {
      node = this.head;
      for (let i = 1; i <= index; i++) {
        node = node.next;
      }
    } else {
      node = this.tail;
      for (let i = this.length - 2; i >= index; i--) {
        node = node.prev;
      }
    }
    return node;
  }

  set(val, index) {
    let setNode = this.get(index);
    if (!setNode) {
      return false;
    }
    setNode.val = val;
    return true;
  }

  insert(val, index) {
    if (index < 0 || index > this.length) return false;
    if (index === 0) return this.shift(val);
    if (index === this.length) return this.push(val);
    const new_node = new Node(val);
    let prevNode = this.get(index - 1);
    new_node.next = prevNode.next;
    prevNode.next.prev = new_node;
    prevNode.next = new_node;
    new_node.prev = prevNode;
    this.length++;
    return true;
  }

  remove(index) {
    if (index < 0 || index >= this.length) return false;
    if (index === 0) return this.unShift();
    if (index === this.length - 1) return !!this.pop();
    let removeNode = this.get(index);
    removeNode.prev.next = removeNode.next;
    removeNode.next.prev = removeNode.prev;
    removeNode.next = null;
    removeNode.prev = null;
    this.length--;
    return true;
  }

  reverse() {
    if (!this.head) return false;
    let node = this.head;
    this.head = this.tail;
    this.tail = node;
    for (let i = 0; i < this.length; i++) {
      let next_node = node.next;
      [node.next, node.prev] = [node.prev, node.next];
      node = next_node;
    }
    return true;
  }
}

let d = new DoublyLinkedList();

File: Python/5_DataStructures/2_DoublyLinkedList.py
class Node(object):
	def __init__(self, val):
		self.val = val
		self.previous = None
		self.next = None
		
class DoublyLinkedList(object):
	def __init__(self):
		self.head = None
		self.tail = None
		self.length = 0

	def printAll(self):
		elements = []
		node = self.head

		while(node):
			elements.append(node.val)
			node = node.next

		return elements
		
	def push(self, val):
		new_node = Node(val)

		if self.length == 0:
			self.head = self.tail = new_node
		else:
			self.tail.next = new_node
			new_node.previous = self.tail
			self.tail = new_node

		self.length += 1
		return True

	def pop(self):
		if self.length == 0: return False
		if self.length == 1: 
			self.head = self.tail = None
		else:
			node = self.tail
			self.tail = node.previous
			self.tail.next = None
			node.previous = None

		self.length -= 1
		return node.val

	def unshift(self, val):
		new_node = Node(val)
		if self.length == 0: 
			self.head = self.tail = new_node
		else:
			new_node.next = self.head
			self.head.previous = new_node
			self.head = new_node

		self.length += 1
		return True

	def shift(self):
		if self.length == 0: return False
		if self.length == 1:
			node = self.head
			self.head = self.tail = None
		else:
			node = self.head
			self.head = node.next
			self.head.previous = None
			node.next = None

		self.length -= 1
		return node

	def get(self, index):
		if index < 0 or index >= self.length: return False
		if index < int(self.length/2):
			node = self.head
			for i in range(index):
				node = node.next
		else:
			node = self.tail
			for i in range(self.length-index - 1):
				node = node.previous

		return node

	def set(self, index, val):
		set_node = self.get(index)
		if not set_node: return False

		set_node.val = val

		return True

	def insert(self, index, val):
		if index < 0 or index > self.length: return False
		if index == 0: return self.unshift(val) != False
		if index == self.length: return self.push(val) != False

		new_node = Node(val)
		previous_node = self.get(index-1)
 
		new_node.next = previous_node.next
		previous_node.next.previous = new_node
		new_node.previous = previous_node
		previous_node.next = new_node

		self.length += 1
		return True

	def remove(self, index):
		if index < 0 or index >= self.length: return False
		if index == 0: return self.shift() != False
		if index == self.length-1: return self.pop() != False

		index_node = self.get(index)

		previous_node = index_node.previous
		next_node = index_node.next
		previous_node.next = next_node
		next_node.previous = previous_node

		self.length -= 1
		return True

	def reverse(self):
		if self.length == 0: return
		self.head, self.tail = self.tail, self.head

		node = self.tail
		for i in range(self.length):
			_next = node.next
			node.next = node.previous
			node.previous = _next

			node = _next

		return True